core/src/software/aws/toolkits/core/utils/Waiter.kt

```diff
@@ -3,14 +3,87 @@
 
 package software.aws.toolkits.core.utils
 
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.runBlocking
 import software.amazon.awssdk.awscore.exception.AwsServiceException
 import java.time.Duration
+import java.time.Instant
+import java.util.concurrent.CompletionException
+import kotlin.math.min
+import kotlin.reflect.KClass
+
+object Waiters {
+    private val LOG = getLogger<Waiters>()
+
+    /**
+     * Generic waiter method to wait for a certain AWS resource to be in a steady status until it reaches in a failed status
+     * or time out for pulling.
+     * @param T The AWS resource type for querying the status from it
+     */
+    fun <T> waitUntil(
+        // The success predicate based on the returned resource
+        succeedOn: (T) -> Boolean = { true },
+        // The fail predicate based on the returned resource. Return an error message if it fails, null otherwise
+        failOn: (T) -> Boolean = { false },
+        exceptionsToStopOn: Set<KClass<out Exception>> = emptySet(),
```

Comment:
it's not immediately clear what this means? shouldn't we stop on all non ignored exceptions?

Reply:
Ya ill document the params

Reply:
And no, things like IOExceptions should be retried, its more like a if we trigger a delete we want to keep retrying until we get a ResourceNotFoundException so we should stop then

Reply:
Cool